add Nodeparam

googincheng opened the pull request:
<img width="1440" alt="截屏2024-08-11 下午6 13 17" src="https://github.com/user-attachments/assets/05128556-e4b0-4782-8557-94bb6b61adbe">

1. node param editor
2. favorite widgets

googincheng committed:
add node param editor

googincheng committed:
Update nodeParamStore.ts

googincheng committed:
fix eslint of Github CI

googincheng committed:
Update nodeParamStore.ts

add ts-expect-error comment

googincheng committed:
Merge branch 'main' into nodeparam

googincheng committed:
fix text input sync to graph canvas

googincheng committed:
opitimize style and catagories

googincheng committed:
change font base size to 12px

googincheng committed:
revert font base size

revert font base size since it would affect browser test

huchenlei commented on src/components/sidebar/tabs/NodeParamSideBarTab.vue:
```diff
@@ -0,0 +1,329 @@
+        ]"
+        :style="{ backgroundColor: node.bgcolor }"
+      >
+        <template v-for="(widget, index) in node.widgets" :key="index">
```
Can you map different widget types to different components with `<component :is="foo"/>`. You can refer to the `SettingGroup` https://github.com/Comfy-Org/ComfyUI_frontend/blob/d5f30be06d8ce4369f12ef22e27779a5cb9b0423/src/components/dialog/content/setting/SettingGroup.vue#L19-L27

huchenlei replied:
BTW, there is an input slider component which is a synced slider with input box that you can reuse: https://github.com/Comfy-Org/ComfyUI_frontend/blob/main/src/components/dialog/content/setting/InputSlider.vue

You can lift it under `common/`

huchenlei commented on src/stores/nodeParamStore.ts:
```diff
@@ -0,0 +1,273 @@
+    return [
+      { name: 'All', count: nodes.value.length },
+      { name: 'Star', count: favCount },
+      {
```
Is this notnull category really needed? What would be the use case?

googincheng replied:
Since many 'Reroute' node would cause the difficulty of finding target node

huchenlei commented on src/stores/nodeParamStore.ts:
```diff
@@ -0,0 +1,273 @@
+    // @ts-expect-error app.graph._nodes is private, there's no export functions to access it
+    const graphNodes = ref([...app.graph._nodes])
+
+    //todo: remove this when app.graph._nodes supports ref()
```
It seems like we are using a very awkward method to sync with litegraph state. I think it is fine for now. However do you have any design on how should make litegraph objects reactive, while maintaining the performance? (For example, we probably don't want to make node position reactive, as that would cause lag when dragging the node)

googincheng replied:
Possible method:
1. easiest way: add callback on node/widget value change, without rewrite the litegraph.js
2. Add a get() set() to litegraph objects, which should be a big rewrite task which may cause unpredictive bug
Love to hear your advice

huchenlei commented on src/stores/nodeParamStore.ts:
```diff
@@ -0,0 +1,273 @@
+    ) // 过滤掉 null 值
+  }
+
+  // const addGalleryImage = (image: string) => {
```
Don't submit commented code.

huchenlei commented on src/components/sidebar/tabs/NodeParamSideBarTab.vue:
```diff
@@ -0,0 +1,329 @@
+<template>
+  <Accordion value="0">
+    <!-- gallery: maybe will merge to other sidepanel -->
```
Don't submit commented code.

huchenlei reviewed (changes requested):
Sorry for the delayed review. I think the main issue is to breakdown the widget to Vue component mapping.

googincheng committed:
Optimize node param sidebar

1、Map nodes widget to <component>
2、Add virtual scroller for hundreds of nodes performance
3、Seperate node and widget since virtual scroller don't support dynamic height element

googincheng commented:
<img width="1440" alt="截屏2024-08-16 下午10 33 10" src="https://github.com/user-attachments/assets/159c0732-f219-4338-9e36-299678b55156">

googincheng committed:
fix eslint

googincheng committed:
fix eslint

googincheng committed:
delete useless code

googincheng committed:
Merge branch 'main' into nodeparam

doctorpangloss commented:
i think you probably want to do exactly the unity blackboard:

https://docs.unity3d.com/Packages/com.unity.shadergraph@12.0/manual/Blackboard.html

https://www.youtube.com/watch?v=TbZYoSu1w8Y

i.e., there's a pinned window with nodes inside of it (a second lightgraph workspace), which you can create unlimited duplicates of inside the main workspace.

the window can contain nodes like ksampler, or more usefully, it would contain primitives. then these become "parameters"

the way you have things now, a node author would have to write a widget for both lightgraph and vue. for example, a color picker. if you just used another lightgraph area, then there's only the one lightgraph widget.

it also isn't that useful to show the nodes in list form.

SirVeggie commented:
Having all nodes listed in the "all" tab makes it very cluttered. I'd rather have it so that I'll never need to look at or use it, and one way to do that would be to allow right clicking on a node widget -> add to param editor. That way you don't always need to scroll through the list, and you can instead just have the "star" tab open at all times.

googincheng commented:
Close PR since this feature still under discussion
